### PR TITLE
remove deprecations from `Domain`

### PR DESCRIPTION
**Pull request by wochinge**
**Proposed changes**:
- fix https://github.com/RasaHQ/rasa/issues/8868
- enforce `Domain` schema
- Exclude the migration guide from this check as the migration guide might contain training data which once was valid but is not valid anymore. E.g. the migration guide for Rasa Open Source 2 defined a form schema which is no longer valid since Rasa Open Source 3.

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: rasa/shared/core/domain.py [hunk 1]
```diff
@@ -702,35 +698,11 @@ def _initialize_forms(
...
-            f"The `forms` section in the domain needs to contain a dictionary. "
-            f"Instead found an object of type '{type(forms)}'.",
-            docs=DOCS_URL_FORMS,
-        )
-
-        return [], {}, []
+        for form_name, form_data in forms.items():
+            if form_data is not None and REQUIRED_SLOTS_KEY not in form_data:
+                forms[form_name] = {REQUIRED_SLOTS_KEY: form_data}
+        # dict with slot mappings
```
Comment (wochinge) [inline rasa/shared/core/domain.py:704]:
```suggestion

```

**Comment by wochinge** [comment]
@m-vdb Is it possible to exclude the migration guide from the docs training data validation? I don't think we should adapt old migration guide items.

**Comment by m-vdb** [comment]
not sure what you're referring to exactly 🤔 we do have [this version by version migration guide](https://rasa.com/docs/rasa/migration-guide)

**Comment by wochinge** [comment]
@m-vdb I am referring to [this failing check](https://github.com/RasaHQ/rasa/pull/9101/checks?check_run_id=3046327128)

**Comment by m-vdb** [comment]
Ah - I think we still need the migration guide though. So it's probably best to move the files closer to the docs (inside the /docs folder for instance), instead of removing the files and the check.

**Comment by wochinge** [comment]
@m-vdb I don't want to remove the migration guide. I just want to exclude the migration guide from the docs check which checks that our documentation only contains valid training data.

**Comment by m-vdb** [comment]
if you exclude the file, then the docs will be broken and won't build

**Comment by m-vdb** [comment]
@wochinge there's no existing mechanism to do that yet, so feel free to update the python test to exclude the migration guide and explain why, and ask my review!

File: tests/docs/test_docs_training_data.py [hunk 2]
```diff
@@ -24,6 +24,13 @@
 
 @pytest.mark.parametrize("mdx_file_path", MDX_DOCS_FILES)
 def test_docs_training_data(mdx_file_path: Path):
+    if mdx_file_path == Path("docs/docs/migration-guide.mdx"):
```
Comment (wochinge) [inline tests/docs/test_docs_training_data.py:27]: @m-vdb That's the part which'd require review by you

Comment (m-vdb) [reply tests/docs/test_docs_training_data.py:27]: a cleaner way could be to remove the path from the `MDX_DOCS_FILES` variable with `MDX_DOCS_FILES.remove(Path("docs/docs/migration-guide.mdx"))` after line 16.

**Review by m-vdb — APPROVED** [review]
sounds good, proposed an alternative for the test

File: rasa/shared/core/domain.py [hunk 3]
```diff
@@ -782,17 +738,6 @@ def num_states(self) -> int:
         """Number of used input states for the action prediction."""
         return len(self.input_states)
 
-    @rasa.shared.utils.common.lazy_property
-    def retrieval_intent_templates(self) -> Dict[Text, List[Dict[Text, Any]]]:
-        """Return only the responses which are defined for retrieval intents."""
-        rasa.shared.utils.io.raise_deprecation_warning(
```
Comment (ancalita) [inline rasa/shared/core/domain.py:788]: Your eagle eyes found a deprecation warning I missed 😅 👍

File: rasa/shared/utils/schemas/domain.yml [hunk 4]
```diff
@@ -44,8 +44,18 @@ mapping:
...
-    type: "any"
+    type: "map"
     required: False
+    mapping:
+      regex;([A-Za-z]+):
+        type: "map"
+        mapping:
+          required_slots:
+            type: "map"
+            required: False
```
Comment (ancalita) [inline rasa/shared/utils/schemas/domain.yml:55]: Should we enforce this to be `required: True` actually? 🤔 
In the docs examples I've seen, `required_slots` is always present, but not sure if giving flexibility to miss this key in forms is welcome or unwelcome.

File: tests/shared/core/test_domain.py [hunk 5]
```diff
@@ -779,16 +771,6 @@ def test_check_domain_sanity_on_invalid_domain():
...
         )
 
-    with pytest.raises(InvalidDomain):
-        Domain(
-            intents={},
-            entities=[],
-            slots=[],
-            responses={},
-            action_names=[],
-            forms=["random_name", "random_name"],
```
Comment (ancalita) [inline tests/shared/core/test_domain.py:789]: Why was this test removed - shouldn't we test that forms as `list` raises `InvalidDomain`?

File: tests/shared/core/test_domain.py [hunk 6]
```diff
@@ -1400,24 +1409,6 @@ def test_ignored_intents_slot_mappings_invalid_domain():
         Domain.from_dict(domain_as_dict)
 
 
-def test_form_with_no_required_slots_keyword():
```
Comment (ancalita) [inline tests/shared/core/test_domain.py:1403]: I think this test should stay, unless you enforce `required_slots` in the schema.

**Review by ancalita — COMMENTED** [review]
Are you planning to open a different PR for this deprecation warning in `rasa/shared/core/slots.py:L317`? It was the only one missed from the issue.
```
"f""{UnfeaturizedSlot.__name__} is deprecated ""
f""and will be removed in Rasa Open Source ""
f""3.0. Please change the type and configure the 'influence_conversation' ""
f""flag for slot '{name}' instead."""
```

File: tests/shared/core/test_domain.py (hunk 5 above)
Comment (wochinge) [reply tests/shared/core/test_domain.py:789]: I dropped it as it's covered by the schema. Can re-add it though.

File: rasa/shared/utils/schemas/domain.yml (hunk 4 above)
Comment (wochinge) [reply rasa/shared/utils/schemas/domain.yml:55]: I put `required: False` so that you can still do 

```
forms:
  my_form:
```

Comment (wochinge) [reply rasa/shared/utils/schemas/domain.yml:55]: (e.g. if all slot mappings are optional)

File: tests/shared/core/test_domain.py (hunk 6 above)
Comment (wochinge) [reply tests/shared/core/test_domain.py:1403]: re-added

File: tests/shared/core/test_domain.py (hunk 5 above)
Comment (wochinge) [reply tests/shared/core/test_domain.py:789]: re-added

**Comment by wochinge** [comment]
> Are you planning to open a different PR for this deprecation warning in rasa/shared/core/slots.py:L317? It was the only one missed from the issue.

Sorry, I've overlooked that one! Thanks the heads up 💯

**Review by ancalita — APPROVED** [review]
💯 
Only pending the resolution of the failed CI check on missing docstring in `CategoricalSlot`.

**Pull request merged**